pkg/models/commandvalue.go

```diff
@@ -30,632 +26,593 @@ const (
 	}
-	err := json.Unmarshal([]byte(cv.stringValue), &value)
-	return value, err
+	value, ok := cv.Value.([]float32)
+	if !ok {
+		errMsg := fmt.Sprintf("failed to transfrom %v to %T", cv.Value, value)
+		return value, errors.NewCommonEdgeX(errors.KindServerError, errMsg, nil)
+	}
+	return value, nil
 }
 
 // Float64Value returns the value in float64 data type, and returns error if the Type is not Float64.
-func (cv *CommandValue) Float64Value() (float64, error) {
+func (cv *CommandValue) Float64Value() (float64, errors.EdgeX) {
 	var value float64
 	if cv.Type != v2.ValueTypeFloat64 {
-		return value, fmt.Errorf("the data type is not %T", value)
+		errMsg := fmt.Sprintf("cannot convert CommandValue of %s to %s", cv.Type, v2.ValueTypeFloat64)
+		return value, errors.NewCommonEdgeX(errors.KindServerError, errMsg, nil)
 	}
-	err := decodeValue(bytes.NewReader(cv.NumericValue), &value)
-	return value, err
+	value, ok := cv.Value.(float64)
+	if !ok {
+		errMsg := fmt.Sprintf("failed to transfrom %v to %T", cv.Value, value)
+		return value, errors.NewCommonEdgeX(errors.KindServerError, errMsg, nil)
+	}
+	return value, nil
 }
 
 // Float64ArrayValue returns the value in an array of float64 type, and returns error if the Type is not Float64Array.
-func (cv *CommandValue) Float64ArrayValue() ([]float64, error) {
+func (cv *CommandValue) Float64ArrayValue() ([]float64, errors.EdgeX) {
 	var value []float64
 	if cv.Type != v2.ValueTypeFloat64Array {
-		return value, fmt.Errorf("the data type is not %T", value)
+		errMsg := fmt.Sprintf("cannot convert %s to %s", cv.Type, v2.ValueTypeFloat64Array)
+		return value, errors.NewCommonEdgeX(errors.KindServerError, errMsg, nil)
 	}
-	err := json.Unmarshal([]byte(cv.stringValue), &value)
-	return value, err
+	value, ok := cv.Value.([]float64)
+	if !ok {
+		errMsg := fmt.Sprintf("failed to transfrom %v to %T", cv.Value, value)
+		return value, errors.NewCommonEdgeX(errors.KindServerError, errMsg, nil)
+	}
+	return value, nil
 }
 
 // BinaryValue returns the value in []byte data type, and returns error if the Type is not Binary.
-func (cv *CommandValue) BinaryValue() ([]byte, error) {
+func (cv *CommandValue) BinaryValue() ([]byte, errors.EdgeX) {
 	var value []byte
 	if cv.Type != v2.ValueTypeBinary {
-		return value, fmt.Errorf("the CommandValue (%s) data type (%v) is not binary", cv.String(), cv.Type)
+		errMsg := fmt.Sprintf("cannot convert %s to %s", cv.Type, v2.ValueTypeBinary)
+		return value, errors.NewCommonEdgeX(errors.KindServerError, errMsg, nil)
+	}
+	value, ok := cv.Value.([]byte)
+	if !ok {
+		errMsg := fmt.Sprintf("failed to transfrom %v to %T", cv.Value, value)
+		return value, errors.NewCommonEdgeX(errors.KindServerError, errMsg, nil)
+	}
+	return value, nil
+}
+
+// Replace checks if the newValue is the same Type with current one by performing
+// type assertion, and replace current value if the assertion succeeded
+func (cv *CommandValue) Replace(newValue interface{}) errors.EdgeX {
```

Comment:
Why/Where is this function needed? Whay not just force use to create a new CommandValue...?

Reply:
this is used if the reading value changed after performing data transformation based on PropertyValue.
It should be `transformer` package's job to ensure newValue is the same valueType with the original one so I'm gonna remove this function.